components/net/fetch/methods.rs

```diff
@@ -486,13 +490,51 @@ fn scheme_fetch(request: &mut Request,
                     Ok(file_path) => {
                         match File::open(file_path.clone()) {
                             Ok(mut file) => {
-                                let mut bytes = vec![];
-                                let _ = file.read_to_end(&mut bytes);
                                 let mime = guess_mime_type(file_path);
 
                                 let mut response = Response::new(url);
-                                *response.body.lock().unwrap() = ResponseBody::Done(bytes);
                                 response.headers.set(ContentType(mime));
+
+                                let (done_sender, done_receiver) = channel();
+                                *done_chan = Some((done_sender.clone(), done_receiver));
+                                *response.body.lock().unwrap() = ResponseBody::Receiving(vec![]);
+
+                                let mut res_body = response.body.clone();
+
+                                let cancellation_listener = context.cancellation_listener.clone();
+
+                                thread::Builder::new().name("fetch file worker thread".to_string()).spawn(move || {
+                                    let mut reader = BufReader::with_capacity(FILE_CHUNK_SIZE, file);
+                                    loop {
+                                        if cancellation_listener.lock().unwrap().cancelled() {
+                                            *res_body.lock().unwrap() = ResponseBody::Done(vec![]);
+                                            let _ = done_sender.send(Data::Cancelled);
+                                            return;
+                                        }
+                                        let length = {
+                                            let mut buffer = reader.fill_buf().unwrap().to_vec();
+                                            let buffer_len = buffer.len();
+                                            if let ResponseBody::Receiving(ref mut body) = *res_body.lock().unwrap() {
+                                                body.extend_from_slice(&buffer);
+                                                let _ = done_sender.send(Data::Payload(buffer.clone()));
```

Comment:
Why is this clone necessary?

Reply:
It isn't, I forgot to remove it because I was doing some testing earlier. It's fixed now.